### DISCUSSION TITLE
does `hardlink`'s `--exclude` prevent descending a directory?

### DISCUSSION THREAD

**Discussion by brianjmurrell** [util-linux/util-linux#3244] [Q&A] [answered]
Does/can `hardlink`'s `--exclude` option actually prevent descending a directory tree or does it just prevent comparison after the entire directory tree has been examined?

I have two directories I want to hardlink where I could save a lot of time and computation by being able to prevent descending one particular subdirectory in each of the two directories.  If I were using `find` for example, I would use the `-prune` action with this particular subdirectory because there will never be hardlinking candidates in it and it can be rather large.

**Comment by karelzak** [comment]
hardlink visits every regular file in the tree, and the --include / --exclude expressions are applied for each file. So, it does not prevent the visit of any subdirectory. The expression is applied to the file names, not to directory names (it internally ignores directories).

**Reply by karelzak** [reply]
I can imagine implementing --exclude-dir, but it will require re-implementing the way it walks the files tree. Currently, it uses the nftw() function from libc.

**Reply by brianjmurrell** [reply]
Isn't `FTW_SKIP_SUBTREE` for this purpose?

**Reply by karelzak** [reply, marked as answer]
Ah, good point. 

I had no clue that nftw() can react to the callback's return value in a smart way  (FTW_ACTIONRETVAL). Yes, it seems we can implement --exclude-subtree easily :-)

**Reply by brianjmurrell** [reply]
> Yes, it seems we can implement --exclude-subtree easily :-)

I thought so to.  I was hoping to find some time soonish to create a PR, but not sure I will find the time before you will (given how simple it is).  So please do go ahead if you find a minute.  Not sure when I will.

If I remember, I will update the _answer_ here with a pointer to a PR/landing commit if/when that happens rather than the answer I did choose above (temporarily).

**Comment by brianjmurrell** [comment]
I wonder if, while you are at it, a `--mount` option utilizing `FTW_MOUNT` would also be useful.

**Comment by karelzak** [comment]
Should be implemented by PR: https://github.com/util-linux/util-linux/pull/3261

**Reply by brianjmurrell** [reply]
Very nice.  I seem to have run into a small usability issue though.  When I execute something like:

```
        /bin/hardlink -n -vv --minimum-size 0 --maximize --respect-dir --respect-name \
	              --exclude-subtree /home --exclude-subtree /var                  \
                      "$SNAPROOT/$1".0/{pc,jenny}
```

I'm getting matches for exclude such as \:

```
Skipped (excluded subtree) /.snapshots/daily.0/pc/usr/include/boost/variant
Skipped (excluded subtree) /.snapshots/daily.0/pc/usr/include/boost/hana/detail/variadic
Skipped (excluded subtree) /.snapshots/daily.0/pc/usr/include/boost/spirit/repository/home
Skipped (excluded subtree) /.snapshots/daily.0/pc/usr/include/boost/spirit/home
```

Should I be using RE anchors such as:
```
        /bin/hardlink -n -vv --minimum-size 0 --maximize --respect-dir --respect-name \
	              --exclude-subtree ^/home --exclude-subtree ^/var                  \
                      "$SNAPROOT/$1".0/{pc,jenny}
```
(i.e. will those work as I have used them above?)

Additionally, should/ do I [need to] further tighten my REs with:

```
        /bin/hardlink -n -vv --minimum-size 0 --maximize --respect-dir --respect-name \
	              --exclude-subtree ^/home/ --exclude-subtree ^/var/                  \
                      "$SNAPROOT/$1".0/{pc,jenny}
```

To avoid skipping things like `/.snapshots/daily.0/pc/homeopathic` and `/.snapshots/daily.0/pc/variant` or would those already not match?

**Comment by karelzak** [comment]
The expression is used with a full path when entering the directory. So, the directory name is the **last item in the path** -- it means it's a string after the last '/'. For example, if I have:
```
/tmp/kzak/homeopathic
/tmp/kzak/home
/tmp/kzak/athome
```
`--exclude-subtree '/home$'` will exclude `/tmp/kzak/home`, but not `/tmp/kzak/homeopathic` or `/tmp/kzak/athome`

It makes sense to use '/' after the directory name only if you want to check the parental directory too. For example `--exclude-subtree '/home/bob$'` to accept `/foo/bob`, but exclude  `/home/bob`.

**Reply by brianjmurrell** [reply]
> The expression is used with a full path when entering the directory. 

So a full path including any prepending directories?  I.e. when hardlinking the two directory trees `/.snapshots/daily.0/{pc,jenny}/` the expression will be matched against `/.snapshots/daily.0/pc/foo` and `/.snapshots/daily.0/jenny/foo`?  If so, and given that [EREs are allowed](https://github.com/util-linux/util-linux/blob/9e2aafe5a493284615a17572c4ead4737bcc66a3/misc-utils/hardlink.c#L273) I guess I can use `(pc|jenny)` in expressions with full paths to specifically match directory names only at specific depths such as:

```
        /bin/hardlink -n -vv --minimum-size 0 --maximize --respect-dir --respect-name \
	              --exclude-subtree "^$SNAPROOT/$1".0/(pc|jenny)/home/ \
	              --exclude-subtree "^$SNAPROOT/$1".0/{pc|jenny}/var/ \
                      "$SNAPROOT/$1".0/{pc,jenny}
```

Correct?

**Reply by karelzak** [reply]
The command line argument "$SNAPROOT/$1".0/{pc,jenny}" is converted into two paths by the shell. The hardlink command sees two independent trees: /.snapshots/daily.0/pc/foo and /.snapshots/daily.0/pc/foo. During the initial stage, when hardlink scans the directories, each path is evaluated separately to filter out relevant files based on user-defined conditions. In the second stage, it works with all accepted files to identify duplicates and create links.

Simple tip (I have to add this to the man page):                                      
                                                                               
If you are unsure about the path used in the regular expression, refer to the verbose output. The path shown in the verbose output is the exact string that  hardlink uses when evaluating the expression (e.g. "Skipped (excluded subtree) **/.snapshots/daily.0/pc/usr/include/boost/variant**").
